src/meta_voting/elections.rs

```
@@ -0,0 +1,263 @@
+            next_index: 0,
+        }
+    }
+
+    pub fn all<'a>(&'a self) -> impl Iterator<Item = MetaElectionHandle> + 'a {
+        self.decided
+            .keys()
+            .cloned()
+            .chain(iter::once(MetaElectionHandle::CURRENT))
+    }
+
+    /// Elections that were already decided by us, but not by all peers.
+    pub fn decided<'a>(&'a self) -> impl Iterator<Item = MetaElectionHandle> + 'a {
+        self.decided.keys().cloned()
+    }
+
+    pub fn meta_votes(
+        &self,
+        handle: MetaElectionHandle,
+        event_hash: &Hash,
+    ) -> Option<&BTreeMap<P, Vec<MetaVote>>> {
+        self.get(handle).and_then(|e| e.meta_votes.get(event_hash))
+    }
+
+    pub(crate) fn round_hashes(
+        &self,
+        handle: MetaElectionHandle,
+        peer_id: &P,
+    ) -> Option<&Vec<RoundHash>> {
+        self.get(handle).and_then(|e| e.round_hashes.get(peer_id))
+    }
+
+    pub fn consensus_history(&self) -> &[Hash] {
+        &self.history
+    }
+
+    /// Creates new election and returns handle of the previous elections.
+    pub fn new_election<'a, I>(&mut self, payload_hash: Hash, voters: I) -> MetaElectionHandle
+    where
+        I: IntoIterator<Item = &'a P>,
+        P: 'a,
+    {
+        let previous = mem::replace(&mut self.current, MetaElection::new(voters));
+        let handle = self.next_handle();
+
+        let _ = self.decided.insert(handle, previous);
+        self.history.push(payload_hash);
+
+        handle
+    }
+
+    /// Mark the given election as decided by the given peer. If there are no more undecided peers,
+    /// the election is removed.
+    pub fn mark_as_decided(&mut self, election: MetaElectionHandle, peer_id: &P) {
+        if let Entry::Occupied(mut entry) = self.decided.entry(election) {
+            let _ = entry.get_mut().undecided_peers.remove(peer_id);
+            if entry.get().undecided_peers.is_empty() {
+                let _ = entry.remove();
+            }
+        }
```

A `log_or_panic!()` in an `else` block here might be useful?  In fact, there are several places in this file where we silently ignore errors.  There are probably some which are fine to ignore, but I'd guess that most places where we're retrieving elections after being passed a handle, it might make sense to use `log_or_panic!()` to cover the error cases?